homework-selects/sql_questions.sql

```diff
@@ -3,55 +3,460 @@
+SELECT (first_name || ' ' || last_name) AS employee_name
+FROM employees
+WHERE salary > (
+    SELECT salary
+    FROM employees
+    WHERE employee_id = 163
+    );
+
 --16. the employee id, employee name (first name and last name ) for all employees who earn more than the average salary.
+SELECT employee_id, (first_name || ' ' || last_name) AS employee_name
+FROM employees
+WHERE salary > (
+    SELECT AVG(salary)
+    FROM employees
+);
+
 --17. the employee name ( first name and last name ), employee id and salary of all employees who report to Payam.
+SELECT (first_name || ' ' || last_name) AS emploee_name, employee_id, salary
+FROM employees
+WHERE manager_id = (
+    SELECT employee_id
+    FROM employees
+    WHERE first_name = 'Payam'
+);
+
 --18. the department number, name ( first name and last name ), job and department name for all employees in the Finance department.
+SELECT department_id, (first_name || ' ' || last_name) AS emploee_name, job_title, department_name
+FROM employees
+    INNER JOIN departments USING (department_id)
+    INNER JOIN jobs USING (job_id)
+WHERE department_name = 'Finance';
+
 --19. all the information of an employee whose id is any of the number 134, 159 and 183.
+SELECT *
+FROM employees
+WHERE employee_id IN (134, 159, 183);
+
 --20. all the information of the employees whose salary is within the range of smallest salary and 2500.
+SELECT *
+FROM employees
+WHERE salary BETWEEN (
+    SELECT MIN(salary)
+    FROM employees
+    ) AND 2500;
+
 --21. all the information of the employees who does not work in those departments where some employees works whose id within the range 100 and 200.
+SELECT *
+      FROM employees
+      WHERE department_id NOT IN (
+          SELECT department_id
+          FROM employees
+                   INNER JOIN departments USING (department_id)
+          WHERE employee_id BETWEEN 100 AND 200
+      );
+
 --22. all the information for those employees whose id is any id who earn the second highest salary.
+SELECT *
+FROM employees
+WHERE salary IN (SELECT MAX(salary)
+                 FROM employees
+                 WHERE salary != (SELECT MAX(salary)
+                                  FROM employees)
+);
+
 --23. the employee name( first name and last name ) and hiredate for all employees in the same department as Clara. Exclude Clara.
+SELECT (first_name || ' ' || last_name) AS employee_name, hire_date
+FROM employees
+WHERE department_id = (
+    SELECT department_id
+    FROM employees
+    WHERE first_name = 'Clara'
+    ) AND first_name != 'Clara';
+
 --24. the employee number and name( first name and last name ) for all employees who work in a department with any employee whose name contains a T.
+SELECT employee_id, (first_name || ' ' || last_name) AS employee_name
+FROM employees
+WHERE department_id in (
+    SELECT department_id
+    FROM employees
+    WHERE (first_name || ' ' || last_name) LIKE '%T%'
```

Comment:
You can specify, that the second query, extracts only distinct values, so your query will run faster because it won't contain duplicates, agree?

Reply:
Agreed. Refactored to use **DISTINCT**

Reply:
I don't think that's the case, because it would have been specified in the condition. I would use both T and t only in case the condition has a smaller t, but vice-versa doesn't work I guess.